crawler.go

```diff
@@ -329,21 +359,65 @@ func (c *Crawler) scanRequestWork(workCh chan chan *http.Request, closeCh chan i
 			spider.reqch <- requestAndChan{req: req, ch: resc}
 			select {
 			case re := <-resc:
-				closeRequest(req)
 				if re.err != nil {
 					c.logf("crawler: send HTTP request got error: %v", re.err)
 				} else {
-					go func(res *http.Response) {
+					go func(clonedReq *http.Request, res *http.Response) {
 						defer closeResponse(res)
 						defer func() {
 							if r := recover(); r != nil {
 								c.logf("crawler: Handler got panic error: %v", r)
 							}
 						}()
+
+						var retryResponseCode bool
+						if len(c.RetryHTTPResponseCodes) > 0 {
+							for _, v := range c.RetryHTTPResponseCodes {
+								if res.StatusCode == v {
+									retryResponseCode = true
+									break
+								}
+							}
+						}
+						url := clonedReq.URL.String()
+						var recrawl bool
+						c.urlNumRetriesMu.Lock()
+						if c.urlNumRetries == nil {
+							c.urlNumRetries = make(map[string]int)
+						}
+						if retryResponseCode {
+							val, _ := c.urlNumRetries[url]
+							if val < c.maxRetries() {
+								c.urlNumRetries[url]++
+								recrawl = true
+							}
+						} else {
+							// Finally, we succeeded! Reset the retry count.
+							c.urlNumRetries[url] = 0
+						}
+						c.urlNumRetriesMu.Unlock()
+
+						if recrawl {
+							timeSleep := c.secondsBetweenRetries()
+							if timeSleepHeader, err := strconv.Atoi(res.Header.Get("Retry-After")); err == nil && timeSleepHeader > 0 {
+								timeSleep = timeSleepHeader
+							}
+							select {
+							case <-clonedReq.Context().Done():
+								c.logf("crawler: aborted because context done")
+							case <-time.After(time.Second * time.Duration(rand.Intn(1+2*timeSleep))):
+								// Randomize the sleeps to spread out retries.
+								c.Crawl(clonedReq)
```

Comment:
This is still recursive, right? Just making sure we're okay with that.

Reply:
Two clarifications:
* If by recursion you mean the function is recursive, answer is no. Crawl is just going to enqueue the URL and other go routines will process
* If by recursion, you mean the queue will build up infinitely if we have failures - two things protect against that -- (a) the max retries logic (b) the enqueue logic has a timeout (of 5s) and if it hits the timeout, it'll crap out.

Reply:
Ah, sorry, my bad. Yeah, this is fine since it only queues it up.